### PR TITLE
ChangeValidator Era Notices, Require Set Keys

### PR DESCRIPTION
**Pull request by maxwolff**
admin, era changing notices such as (set_yield, set_supply_cap, change_authority)  now which bump the notice era and set a `NoticeHold` on any new notices being signed until they are signed. The notices are added to`PendingEraNotices`, if they are executed they are removed upon `on_initialize`.

* adds `shouldEndSession`, which is periodic unless an era change is ocurring
* tests the `author_rotateKeys` => `session_setKeys` => `changeAuthorities` flow, and make sure changeAuthorities fails if account does not have grandpa/aura keys set
* goes with this 1 line substrate change https://github.com/compound-finance/substrate/pull/27

### CODE REVIEW

File: integration/__tests__/gov_test.js [hunk 1]
```diff
@@ -31,8 +32,8 @@ buildScenarios('Gov Scenarios', gov_scen_info, [
     scenario: async ({ ctx, chain, starport, validators }) => {
       const alice = validators.validatorInfoMap.alice;
       const alice_account_id = alice.aura_key;
-      const newAuthsRaw = [[ctx.actors.keyring.decodeAddress(alice_account_id), { eth_address: alice.eth_account }]];
-      let extrinsic = ctx.api().tx.cash.changeAuthorities(newAuthsRaw);
+      const newAuthsRaw = [{ substrate_id: ctx.actors.keyring.decodeAddress(alice_account_id), eth_address: alice.eth_account }];
```
Comment (maxwolff) [inline integration/__tests__/gov_test.js]: format changes in a prev pr broke this test

File: pallets/cash/src/core.rs [hunk 2]
```diff
@@ -335,15 +335,31 @@ pub fn apply_chain_event_internal<T: Config>(event: ChainLogEvent) -> Result<(),
...
+    should_increment_era: bool,
     notice_fn: &dyn Fn(NoticeId, ChainHash) -> Result<Notice, Reason>,
 ) -> Result<(), Reason> {
     // XXX this cannot fail, should not return result
     let (latest_notice_id, parent_hash) =
         LatestNotice::get(chain_id).unwrap_or((NoticeId(0, 0), chain_id.zero_hash()));
-    let notice_id = latest_notice_id.seq();
+
+    let notice_id = if should_increment_era {
+        require!(
```
Comment (jflatow) [inline pallets/cash/src/core.rs]: So we return an error on the extrinsic if there are any pending era notices to be signed? Need to think about this, didn't we ideally want to continue accepting transactions, just not signing them yet?

**Review by jflatow — COMMENTED** [review]
Great stuff

File: pallets/cash/src/params.rs [hunk 3]
```diff
@@ -19,3 +19,10 @@ pub const TRANSFER_FEE: Quantity = Quantity::from_nominal("0.01", CASH);
 
 /// Number of blocks between HTTP requests from offchain workers to open oracle price feed.
 pub const ORACLE_POLL_INTERVAL_BLOCKS: u32 = 10;
+
+/// The number of blocks to wait before switching sessions
+#[cfg(debug_assertions)]
+pub const SESSION_PERIOD: u32 = 2;
```
Comment (maxwolff) [inline pallets/cash/src/params.rs:25]: just noticed we set this to 50 before, makes integration tests go slower, what was the reason behindthis?

Comment (coburncoburn) [reply pallets/cash/src/params.rs:25]: can be 2 in test, but running live it is really spammy to be changing sessions every 2 blocks

File: pallets/cash/src/lib.rs [hunk 4]
```diff
@@ -329,6 +359,53 @@ impl<T: Config> pallet_session::SessionManager<SubstrateId> for Module<T> {
...
+        } else {
+            // no era changes pending, periodic
+            let period: T::BlockNumber = <T>::BlockNumber::from(params::SESSION_PERIOD as u32);
+            (now % period) == <T>::BlockNumber::from(0 as u32)
+        }
+    }
+}
+
+impl<T: Config> frame_support::traits::EstimateNextSessionRotation<T::BlockNumber> for Module<T> {
+    fn estimate_next_session_rotation(now: T::BlockNumber) -> Option<T::BlockNumber> {
```
Comment (coburncoburn) [inline pallets/cash/src/lib.rs:450]: is this just part of the interface, or used somehow?

Comment (maxwolff) [reply pallets/cash/src/lib.rs:450]: its just part of the interface, not used for anything afaict

File: pallets/cash/src/core.rs [hunk 5]
```diff
@@ -1381,6 +1399,14 @@ pub fn on_initialize_internal<T: Config>(
     LastMinerSharePrincipal::put(miner_share_principal);
     LastYieldTimestamp::put(now);
 
+    // undo the hold placed on notice signing during era changes
+    for (chain_id, notice_id) in PendingEraNotices::iter() {
```
Comment (coburncoburn) [inline pallets/cash/src/core.rs]: on initialize internal is getting a bit large

File: pallets/cash/src/internal/change_validators.rs [hunk 6]
```diff
@@ -1,14 +1,43 @@
...
+    chains::{ChainHash, ChainId},
+    core::dispatch_notice_internal,
+    notices::{ChangeAuthorityNotice, Notice},
+    reason::Reason,
+    require,
+    types::ValidatorKeys,
+    Config, Event, Module, NextValidators, SessionInterface,
+};
 
 pub fn change_validators<T: Config>(validators: Vec<ValidatorKeys>) -> Result<(), Reason> {
```
Comment (coburncoburn) [inline pallets/cash/src/internal/change_validators.rs:13]: who calls change_validators?

Comment (maxwolff) [reply pallets/cash/src/internal/change_validators.rs:13]: called here, must be root. so either sudo or by gov. https://github.com/compound-finance/compound-chain/blob/e17c2bad5ae7532ffe72f8ea9a1334b4e8fd3eac/pallets/cash/src/lib.rs#L447

File: integration/util/substrate.js [hunk 7]
```diff
@@ -74,6 +74,15 @@ function waitForEvent(api, pallet, method, opts = {}) {
   return promise;
 }
 
+// TODO
```
Comment (hayesgm) [inline integration/util/substrate.js]: You can now use `sendAndWaitForEvents(call, api, { sender: signer })`

File: pallets/cash/src/internal/notices.rs [hunk 8]
```diff
@@ -27,30 +27,37 @@ pub fn handle_notice_invoked<T: Config>(
...
-                        .ok_or(Reason::NoticeMissing(chain_id, notice_id))?;
-                    let signature: ChainSignature = notice.sign_notice()?;
-                    log!("Posting Signature for [{},{}]", notice_id.0, notice_id.1);
-                    let call = <Call<T>>::publish_signature(chain_id, notice_id, signature);
-
-                    // TODO: Do we want to short-circuit on an error here?
-                    let _res =
-                        SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into())
-                            .map_err(|()| Reason::FailedToSubmitExtrinsic);
+        let mut notice_hold = false;
```
Comment (hayesgm) [inline pallets/cash/src/internal/notices.rs]: Why not just take `NoticeHold` and then compare the notice ids (making sure that notices implement ordering?)

File: pallets/cash/src/lib.rs [hunk 9]
```diff
@@ -175,6 +174,15 @@ decl_storage! {
         /// The most recent notice emitted for a given chain
         LatestNotice get(fn latest_notice_id): map hasher(blake2_128_concat) ChainId => Option<(NoticeId, ChainHash)>;
 
+        // Notices that change the starport era. All pending era notices must be signed before new regular notices can be signed
```
Comment (hayesgm) [inline pallets/cash/src/lib.rs]: Can't notice hold just handle this semi-magically?

File: pallets/cash/src/internal/notices.rs (hunk 8 above)
Comment (maxwolff) [reply pallets/cash/src/internal/notices.rs]: not sure i understand

File: pallets/cash/src/core.rs [hunk 10]
```diff
@@ -374,15 +374,26 @@ pub fn apply_chain_event_internal<T: Config>(event: ChainLogEvent) -> Result<(),
...
+    let notice_id = if should_increment_era {
+        require!(NoticeHold::iter().count() == 0, Reason::PendingEraNotice);
+        latest_notice_id.seq_era()
+    } else {
+        latest_notice_id.seq()
+    };
 
     let notice = notice_fn(notice_id, parent_hash)?; // XXX fixme cannot fail
 
+    if should_increment_era {
```
Comment (hayesgm) [inline pallets/cash/src/core.rs:393]: Is this always true? Can we have new eras without a hold? I feel like this is a separate var "should hold"?

File: pallets/cash/src/lib.rs [hunk 11]
```diff
@@ -329,6 +353,111 @@ impl<T: Config> pallet_session::SessionManager<SubstrateId> for Module<T> {
...
+        a_set.insert(v);
+    }
+
+    let mut b_set = BTreeSet::<T>::new();
+    for v in b {
+        b_set.insert(v);
+    }
+
+    let count = a_set.intersection(&b_set).into_iter().count();
+    println!("a={}, b={}, count={}", a_count, b_count, count);
```
Comment (hayesgm) [inline pallets/cash/src/lib.rs:388]: I think my recent patch fixes this (the `recover_user_address` versus `recover_address` fix)

File: pallets/cash/src/lib.rs [hunk 12]
```diff
@@ -329,6 +353,111 @@ impl<T: Config> pallet_session::SessionManager<SubstrateId> for Module<T> {
...
+                true
+            } else {
+                println!("should_end_session=false[notices_held]");
+                false
+            }
+        } else {
+            // no era changes pending, periodic
+            let period: T::BlockNumber = <T>::BlockNumber::from(params::SESSION_PERIOD as u32);
+            let is_new_period = (now % period) == <T>::BlockNumber::from(0 as u32);
+            println!(
```
Comment (hayesgm) [inline pallets/cash/src/lib.rs:440]: We might not want to print every time, but it really helps while we're testing to have this. In the future this probably should, at the least, only show when it's `true` for the periodic check.

**Review by hayesgm — APPROVED** [review]
Looks _excellent_, only question is around whether we should always hold on new eras.

**Pull request merged**